File: pract9/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <set>
#include <string>

using namespace std;

// Check if a token is a literal
bool isLiteral(const string &token) {
    return token.size() > 2 && token.substr(0, 2) == "='";
}

int main() {
    ifstream infile("input.asm");  // Open the asm file
    if (!infile.is_open()) {
        cerr << "Error: Could not open file." << endl;
        return 1;
    }

    string line;
    vector<string> literalTable;
    set<string> literalSet;
    vector<int> poolTable;
    int literalIndex = 0;

    while (getline(infile, line)) {
        stringstream ss(line);
        string token;
        while (ss >> token) {
            if (isLiteral(token)) {
                if (literalSet.find(token) == literalSet.end()) {
                    literalSet.insert(token);
                    literalTable.push_back(token);
                }
            } else if (token == "LTORG" || token == "END") {
                if (literalIndex < literalTable.size()) {
                    poolTable.push_back(literalIndex);
                    literalIndex = literalTable.size(); // update literal pool index
                }
            }
        }
    }

    infile.close();

    // Output Pool Table
    cout << "POOL TABLE" << endl;
    for (int i = 0; i < poolTable.size(); ++i) {
        cout << "#" << (i + 1) << " -> " << poolTable[i] << endl;
    }

    return 0;
}

File: pract9/solution.py
import os

# Check if a token is a literal
def is_literal(token):
    return len(token) > 2 and token.startswith("='")

def main():
    if not os.path.isfile("input.asm"):
        print("Error: Could not open file.")
        return 1

    literal_table = []
    literal_set = set()
    pool_table = []
    literal_index = 0

    with open("input.asm", "r") as infile:
        for line in infile:
            tokens = line.split()
            for token in tokens:
                if is_literal(token):
                    if token not in literal_set:
                        literal_set.add(token)
                        literal_table.append(token)
                elif token in ["LTORG", "END"]:
                    if literal_index < len(literal_table):
                        pool_table.append(literal_index)
                        literal_index = len(literal_table)  # update literal pool index

    # Output Pool Table
    print("POOL TABLE")
    for i in range(len(pool_table)):
        print(f"#{i + 1} -> {pool_table[i]}")

if __name__ == "__main__":
    main()

